### PR TITLE
Fot container

### PR DESCRIPTION
**Pull request by anurag6**
DHCP reflection doesnt work yet, tinkering with the ACL

### CODE REVIEW

File: etc/FAUCET_VERSION [hunk 1]
```diff
@@ -1 +1 @@
-1.9.48
+1.9.49
```
[the remark this answers is not in the document]
Comment (anurag6) [reply etc/FAUCET_VERSION]: Reverted in cleanup.

File: bin/run_controller [hunk 2]
```diff
@@ -32,6 +32,7 @@ fi
 if [ -n "$1" ]; then
     CONTAINER=$CONTAINER-$1
     INSTDIR=$INSTDIR/forch-controller-$1
+    INSTDIR2=$INSTDIR/faucet_br
```
[the remark this answers is not in the document]
Comment (anurag6) [reply bin/run_controller]: Changed.

File: bin/run_controller [hunk 3]
```diff
@@ -106,6 +109,8 @@ docker_envs="\
         --env FAUCET_RYU_CONF=/etc/faucet/ryu.conf"
 
 docker_volumes+="\
+        -v $INSTDIR2/faucet:/etc/faucet_br \
```
[the remark this answers is not in the document]
Comment (anurag6) [reply bin/run_controller]: Changed.

File: bin/setup_stack [hunk 4]
```diff
@@ -116,7 +116,11 @@ for index in 1; do
     cp -r topo/$mode/faucet inst/$container/
     docker kill $container || true
     if [ -n "$fot" ]; then
+        mkdir -p inst/$container/faucet_br/faucet
+        cp -r topo/$mode/faucet_br/* inst/$container/faucet_br/faucet/
+        ls inst/$container/faucet_br/faucet/
```
Comment (grafnu) [inline bin/setup_stack:121]: remove extra debugging for now (the ls here and two lines below)

File: bin/run_controller [hunk 5]
```diff
@@ -150,11 +152,11 @@ if [ $IMAGE == "forch/controller" ] && [[ $faucet_id =~ ^[0-9]+$ ]]; then
     data_intf_name=data_faucet_$faucet_id
     sudo ip link add $data_intf_name type veth peer name data0 netns $pid
     sudo ip link set $data_intf_name up
-    docker exec $CONTAINER ip link add link data0 name data0.272 type vlan id 272
-    docker exec $CONTAINER ip addr add 192.168.2.$faucet_id dev data0.272
+    docker exec $CONTAINER ip addr add 192.168.2.$faucet_id dev data0
```
Comment (grafnu) [inline bin/run_controller:155]: Most if this "if image=controller" section should be moved into the docker file startup script, not as individual docker exec commands (this includes some things that were already here before this PR).

File: bin/run_controller [hunk 6]
```diff
@@ -150,11 +152,11 @@ if [ $IMAGE == "forch/controller" ] && [[ $faucet_id =~ ^[0-9]+$ ]]; then
...
     docker exec $CONTAINER ip link set data0 up
-    docker exec $CONTAINER ip link set data0.272 up
-    docker exec $CONTAINER ip route add 192.168.0.0/16 dev data0.272
+    docker exec $CONTAINER /usr/share/openvswitch/scripts/ovs-ctl start
+    docker cp misc/controller_go $CONTAINER:/usr/bin/controller_go 
+    docker exec $CONTAINER /usr/bin/controller_go
```
Comment (grafnu) [inline bin/run_controller:159]: controller_go should be this last thing in the Dockerfile, instead of faucet_go

File: bin/run_controller (hunk 5 above)
Comment (anurag6) [reply bin/run_controller:153]: I am leaving in the commands linked to data0

File: bin/setup_stack (hunk 4 above)
Comment (anurag6) [reply bin/setup_stack]: Changed.

File: bin/run_controller (hunk 6 above)
Comment (anurag6) [reply bin/run_controller]: Changed. faucet_go isnt called any more.

File: misc/controller_go [hunk 7]
```diff
@@ -0,0 +1,31 @@
+#!/bin/sh -ex
```
Comment (grafnu) [inline misc/controller_go:1]: /bin/bash  (just what we do everywhere)

File: misc/controller_go [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+sudo /usr/share/openvswitch/scripts/ovs-ctl start
+sudo ovs-vsctl --if-exists del-br br-fot \
+    -- add-br br-fot \
+    -- set-fail-mode br-fot secure \
+    -- set bridge br-fot other_config:datapath-id=0xfa
+echo Adding fot bridge controller
+sudo ovs-vsctl set-controller br-fot tcp:127.0.0.1:3984
+/usr/bin/gauge_go &
+/usr/bin/forch &
+sudo -u kronkiteman \
```
Comment (grafnu) [inline misc/controller_go:26]: run this in the background for consistency, then either add a "wait" or "tail -f /dev/null" at the end to block.

File: bin/run_controller [hunk 9]
```diff
@@ -144,17 +146,14 @@ cid=$(docker run -d --privileged --name $CONTAINER \
...
     sudo ip link add $data_intf_name type veth peer name data0 netns $pid
     sudo ip link set $data_intf_name up
-    docker exec $CONTAINER ip link add link data0 name data0.272 type vlan id 272
-    docker exec $CONTAINER ip addr add 192.168.2.$faucet_id dev data0.272
+    docker exec $CONTAINER ip addr add 192.168.2.$faucet_id dev data0
     docker exec $CONTAINER ip link set data0 up
-    docker exec $CONTAINER ip link set data0.272 up
-    docker exec $CONTAINER ip route add 192.168.0.0/16 dev data0.272
+    echo Adding fot bridge port
+    docker exec $CONTAINER bash -c "ovs-vsctl add-port br-fot data0 -- set interface data0 ofport_request=1; ifconfig data0 up"
```
Comment (grafnu) [inline bin/run_controller:156]: This part should really go inside the container.  Although it is *triggered* by an external event (data0 appearing), it's not dependent on it and shouldn't be exposed.  One solution would be to add a "while data0 not configured" wait loop (see etc/start_faux for an example) or the docker exec call here would directly execute a "data0_ready" script inside of the container.

The reason is encapsulation. What you've done here is basically exposing container internals externally.

**Review by grafnu — COMMENTED** [review]
almost there... just a few minor tweaks...

File: misc/controller_go (hunk 7 above)
Comment (anurag6) [reply misc/controller_go]: Changed.

File: misc/controller_go (hunk 8 above)
Comment (anurag6) [reply misc/controller_go:26]: Made the changes.

File: bin/run_controller (hunk 9 above)
Comment (anurag6) [reply bin/run_controller]: Makes sense. Added it to controller_go

**Review by grafnu — APPROVED** [review]

**Pull request merged**